synapse/cryotank.py

```diff
@@ -571,3 +588,609 @@ def init(self, name, conf=None, timeout=None):
+            if prop is None or prop == idx.propname:
+                self.asleep[iid] = True
+
+    def resumeIndex(self, prop):
+        '''
+        Undo a pauseIndex.
+        Args:
+            prop: (Optional[str]):  the index to start indexing, or if None, indicate to resume all indices
+        Returns:
+            None
+        '''
+        for iid, idx in self.indices.items():
+            if prop is None or prop == idx.propname:
+                self.asleep[iid] = False
+
+    def markDeleteComplete(self, iid: int) -> None:
+        self.deleting.remove(iid)
+        self.persist()
+
+# Encodes a little endian 64-bit integer
+_Int64le = struct.Struct('<Q')
+
+def _iid_en(iid):
+    return _Int64le.pack(iid)
+
+# Decodes a little endian 64-bit integer
+def _iid_un(iid):
+    return _Int64le.unpack(iid)[0]
+
+def _inWorker(callback):
+    '''
+    Gives the decorated function to the worker to run in his thread.
+
+    (Just like inpool for the worker)
+    '''
+    def wrap(self, *args, **kwargs):
+        with s_threads.RetnWait() as retn:
+            self._workq.put((retn, callback, (self, ) + args, kwargs))
+            succ, rv = retn.wait(timeout=self.MAX_WAIT_S)
+            if succ:
+                if isinstance(rv, Exception):
+                    raise rv
+                return rv
+            raise s_exc.TimeOut()
+
+    return wrap
+
+class CryoTankIndexer:
+    '''
+    Manages indexing of a single cryotank's records
+
+    This implements a lazy indexer that indexes a cryotank in a separate thread.
+
+    Cryotank entries are msgpack-encoded json-compatible dictionaries with arbitrary nesting.  An index consists of a
+    property name, one or more datapaths (i.e. what field out of the entry), and a synapse type.   The type specifies
+    the function that normalizes the output of the datapath query into a string or integer.
+
+    Indices can be added and deleted asynchronously from the indexing thread via CryotankIndexer.addIndex and
+    CryotankIndexer.delIndex.
+
+    Indexes can be queried with normValuByPropVal, normRecordsByPropVal, rawRecordsByPropVal.
+
+    To harmonize with LMDB requirements, writing only occurs on the worker thread, while reading indices takes place in
+    the caller's thread.  Both reading and writing index metadata (that is, information about which indices are
+    running) take place on the worker's thread.
+
+    N.B. The indexer cannot detect when a type has changed from underneath itself.   Operators must explicitly delete
+    and re-add the index to avoid mixed normalized data.
+    '''
+    MAX_WAIT_S = 10
+
+    def __init__(self, cryotank):
+        '''
+        Create an indexer.
+
+        Args:
+            cryotank: the cryotank to index
+        Returns:
+            None
+        '''
+        self.cryotank = cryotank
+        ebus = cryotank
+        self._going_down = False
+        self._worker = threading.Thread(target=self._workerloop, name='CryoTankIndexer')
```

Comment:
we have a ``worker()`` helper in ``s_threads`` which makes and starts a worker thread - we could probably use him after we have setup our handlers and all other variables.

Reply:
+1000 for naming your thread